src/compiler/crystal/tools/typer.cr

```diff
@@ -0,0 +1,526 @@
+module Crystal
+  class SourceTyper
```

Comment:
In the [source-typer](https://github.com/Vici37/cr-source-typer) repo, a lot of these classes / structs are broken into their own files. Precedence in the compiler seems to be to put a lot of these smaller bits into a single file if they can all be more or less self contained.